packages/muelu/src/Transfers/Smoothed-Aggregation/MueLu_SaPFactory_kokkos_def.hpp

```diff
@@ -360,6 +360,207 @@ struct constraintKernel {
+             ind = sortInds(i);
+             LO j = i;
+             
+             if ( Kokkos::ArithTraits<SC>::real(sortVals(sortInds(i))) < Kokkos::ArithTraits<SC>::real(sortVals(sortInds(0))) ){
+               for ( ; j > 0; --j) sortInds(j) = sortInds(j - 1);
+  
+               sortInds(0) = ind;
+             } else {
+               for ( ; Kokkos::ArithTraits<SC>::real(sortVals(ind)) < Kokkos::ArithTraits<SC>::real(sortVals(sortInds(j-1))); --j) sortInds(j) = sortInds(j-1);
+  
+               sortInds(j) = ind;
+             }
+           }
+  
+  
+           for (LO i = 0; i < (LO) nnz; i++) origSorted(rowIdx, i) = values(rowPtr(rowIdx) + inds(rowIdx, i)); //values is no longer used
+           // find entry in origSorted just to the right of the leftBound
+           closestToLeftBound = 0;
+           while ((closestToLeftBound < (LO) nnz) && (Kokkos::ArithTraits<SC>::real(origSorted(rowIdx, closestToLeftBound)) <= Kokkos::ArithTraits<SC>::real(leftBound))) closestToLeftBound++;
+  
+           // find entry in origSorted just to the right of the rghtBound
+           closestToRghtBound = closestToLeftBound;
+           while ((closestToRghtBound < (LO) nnz) && (Kokkos::ArithTraits<SC>::real(origSorted(rowIdx, closestToRghtBound)) <= Kokkos::ArithTraits<SC>::real(rghtBound))) closestToRghtBound++;
+  
+           // compute distance between closestToLeftBound and the left bound and the 
+           // distance between closestToRghtBound and the right bound. 
+        
+           closestToLeftBoundDist = origSorted(rowIdx, closestToLeftBound) - leftBound;
+           if (closestToRghtBound == (LO) nnz) closestToRghtBoundDist= aBigNumber;
+           else                                closestToRghtBoundDist= origSorted(rowIdx, closestToRghtBound) - rghtBound;
+  
+           // compute how far the rowSum is off from the target row sum taking into account
+           // numbers that have been shifted to satisfy bound constraint
+  
+           rowSumDeviation = leftBound*((SC)closestToLeftBound) + ((SC)(nnz-closestToRghtBound))*rghtBound - rsumTarget; 
+           for (LO i=closestToLeftBound; i < closestToRghtBound; i++) rowSumDeviation += origSorted(rowIdx, i);
+  
+           // the code that follow after this if statement assumes that rowSumDeviation is positive. If this
+           // is not the case, flip the signs of everything so that rowSumDeviation is now positive. 
+           // Later we will flip the data back to its original form.
+           if (Kokkos::ArithTraits<SC>::real(rowSumDeviation) < Kokkos::ArithTraits<SC>::real(Kokkos::ArithTraits<SC>::zero())) {
+             flipped = true;
+             temp = leftBound; leftBound = -rghtBound; rghtBound = temp; 
+         
+             /* flip sign of origSorted and reverse ordering so that the negative version is sorted */
+         
+             if ((nnz%2) == 1) origSorted(rowIdx, (nnz/2)  ) =  -origSorted(rowIdx, (nnz/2)  );
```

Comment:
The above is pretty bad for GPU performance... you can do bit shifting instead if you feel brave.

Reply:
@lucbv Is this something that @pohm01 could do later in a separate PR?

Reply:
Added a comment noting that this is bad for GPU performance, but I feel switching to bit shifting is beyond the scope of this PR. This can be addressed later if this function begins to see heavy use.